tests/unit/test_concurrency.py

```diff
@@ -12,140 +12,63 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import collections
 import logging
+import os
+
+import pytest
 
 try:
     from unittest import mock
 except ImportError:  # pragma: NO PY3 COVER
     import mock
 
 from google.cloud.ndb import _cache
-from google.cloud.ndb import _eventloop
 from google.cloud.ndb import global_cache as global_cache_module
 from google.cloud.ndb import tasklets
-from google.cloud.ndb import utils
 
+from . import orchestrate
 
 log = logging.getLogger(__name__)
 
 
-class Delay(object):
-    """A tasklet wrapper which delays the return of a tasklet.
-
-    Used to orchestrate timing of events in async code to test particular scenarios
-    involving concurrency. Use with `mock.patch` to replace particular tasklets with
-    wrapped versions. When those tasklets are called, they will execute and then the
-    wrapper will hang on to the result until :meth:`Delay.advance()` is called, at which
-    time the tasklet's caller will receive the result.
-
-    Args:
-        wrapped (tasklets.Tasklet): The tasklet to be delayed.
-    """
-
-    def __init__(self, wrapped):
-        self.wrapped = wrapped
-        self.info = "Delay {}".format(self.wrapped.__name__)
-        self._futures = collections.deque()
-
-    @tasklets.tasklet
-    def __call__(self, *args, **kwargs):
-        future = tasklets.Future(self.info)
-        self._futures.append(future)
+def cache_factories():
+    yield global_cache_module._InProcessGlobalCache
 
-        result = yield self.wrapped(*args, **kwargs)
-        yield future
-        raise tasklets.Return(result)
+    def redis_cache():
+        return global_cache_module.RedisCache.from_environment()
 
-    def advance(self):
-        """Allow a call to the wrapper to proceed.
+    if os.environ.get("REDIS_CACHE_URL"):
+        yield redis_cache
 
-        Calls are advanced in the order in which they were orignally made.
-        """
-        self._futures.popleft().set_result(None)
+    def memcache_cache():
+        return global_cache_module.MemcacheCache.from_environment()
 
+    if os.environ.get("MEMCACHED_HOSTS"):
+        yield global_cache_module.MemcacheCache.from_environment
```

Comment:
Can't be reached because of the `return` above.

Reply:
I think you might have misread this one. The return is from a locally defined function.

Reply:
Yup.  Hard to track the whitespace across diffs, especially with interleaved comments.